File: q58.py
"""
58. Length of Last Word

Given a string s consisting of some words separated by some
number of spaces, return the length of the last word in the
string.

A word is a maximal substring consisting of non-space
characters only.

https://leetcode.com/problems/length-of-last-word/
"""


class Solution:
    def lengthOfLastWord(self, s: str) -> int:
        i = len(s) - 1
        acc = 0
        word_matched = False
        while i >= 0:
            if s[i] == " ":
                if word_matched:
                    break
            else:
                if not word_matched:
                    word_matched = True
                acc += 1
            i -= 1
        return acc


print(Solution().lengthOfLastWord("   fly me   to   the moon  "))

File: q58.cpp
/**
 * 58. Length of Last Word
 * https://leetcode.com/problems/length-of-last-word
 */

#include "my_util.hpp"

using namespace std;

class Solution {
public:
    int lengthOfLastWord(string s) {
        int lastEnd = s.size();
        while (lastEnd > 0 && s[lastEnd - 1] == ' ') {
            --lastEnd;
        }

        if (lastEnd <= 0) {
            return s.size();  // all word
        }

        int lastStart = lastEnd - 1;
        while (lastStart >= 0 && s[lastStart] != ' ') {
            --lastStart;
        }

        if (lastStart < 0) {
            return lastEnd;
        }

        return lastEnd - lastStart - 1;
    }
};

int main() {
    print_v(Solution().lengthOfLastWord("hello this is a test"));
    print_v(Solution().lengthOfLastWord("hello this is a smalltest"));
    print_v(Solution().lengthOfLastWord("anothertest"));
    print_v(Solution().lengthOfLastWord("   fly me   to   the moon  "));
    print_v(Solution().lengthOfLastWord("a "));
}
